### PR TITLE
tool: HTML String Check

### PR DESCRIPTION
**Pull request by BenAndrew-Kainos**
String checker for finding HTML strings that are missing from lite_content.

Internal: https://github.com/uktrade/lite-internal-frontend/pull/274

### CODE REVIEW

File: templates/check_html_strings.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import importlib
+import os
+import re
+
+LCS_REGEX = "{% lcs '(.*)' %}"
```
Comment (stephanieboyd) [inline templates/check_html_strings.py]: Let's call this `lcs_pattern`

File: templates/check_html_strings.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def find_occurance(module, string):
+    """
+    Check that the given string variable i.e. CASES.Page.TITLE is found within a given module (strings.py).
```
Comment (stephanieboyd) [inline templates/check_html_strings.py]: Backtick the variable name

File: templates/check_html_strings.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    Checks each section of the path is found inside the module defined in the last iteration
+    :param module: Python module to search
+    :param string: Path to find (split into sections)
+    :return: bool of whether it was found
```
Comment (stephanieboyd) [inline templates/check_html_strings.py]: just write true/false here

File: .circleci/config.yml [hunk 4]
```diff
@@ -77,3 +77,7 @@ jobs:
           command: |
             pipenv install bandit
             pipenv run bandit -r .
+
+      - run:
+          name: Run String checker
```
Comment (stephanieboyd) [inline .circleci/config.yml]: `Run content string check`

File: templates/check_html_strings.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    strings_module = importlib.import_module(f"lite_content.{strings_package_name}.strings")
+
+    # Check all strings are found
+    lcs_strings = get_all_lcs_strings(templates_folder)
+    for string in lcs_strings:
+        if not find_occurance(strings_module, string.split(".")):
+            not_found.append(string)
+
+    if not_found:
+        raise Exception(f"The following strings couldn't be found: {not_found}")
```
Comment (stephanieboyd) [inline templates/check_html_strings.py]: I've left a lot of these comments on the one I did review - I don't think we should be raising an exception - just the script output is strings or no strings.

File: templates/check_html_strings.py (hunk 3 above)
Comment (BenAndrew-Kainos) [reply templates/check_html_strings.py]: Done

File: templates/check_html_strings.py (hunk 2 above)
Comment (BenAndrew-Kainos) [reply templates/check_html_strings.py]: Done

File: templates/check_html_strings.py (hunk 1 above)
Comment (BenAndrew-Kainos) [reply templates/check_html_strings.py]: Done

File: .circleci/config.yml (hunk 4 above)
Comment (BenAndrew-Kainos) [reply .circleci/config.yml]: done

File: templates/check_html_strings.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    """
+    for path_section in string:
+        module = module.__dict__.get(path_section)
+        if not module:
+            return False
+
+    return True
+
+
+def get_strings_package(base_dir):
```
Comment (stephanieboyd) [inline templates/check_html_strings.py]: there is a function for this, in os package.
```>>> os.path.basename("/Users/user/lite-internal-frontend")
'lite-internal-frontend'````

File: templates/check_html_strings.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+def get_all_lcs_strings(templates_folder):
+    """
+    Get all LCS strings from the HTML files in the given folder
+    :param templates_folder: string path for the HTML folders
+    :return: set of found LCS strings
+    """
+    strings = set()
+
+    for root, _, files in os.walk(templates_folder):
+        for file in files:
```
Comment (stephanieboyd) [inline templates/check_html_strings.py]: Can you change everywhere you've used the file keyword? Call it `file_name`

File: templates/check_html_strings.py (hunk 6 above)
Comment (BenAndrew-Kainos) [reply templates/check_html_strings.py]: Done

File: templates/check_html_strings.py (hunk 7 above)
Comment (BenAndrew-Kainos) [reply templates/check_html_strings.py]: Done

File: templates/check_html_strings.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    # Check all strings are found
+    lcs_strings = get_all_lcs_strings(templates_folder)
+    for string in lcs_strings:
+        if not find_occurance(strings_module, string.split(".")):
+            not_found.append(string)
+
+    if not_found:
+        raise Exception(f"The following strings couldn't be found: {not_found}")
+    else:
+        print("All strings found")
```
Comment (stephanieboyd) [inline templates/check_html_strings.py]: "No unused strings found"?

File: templates/check_html_strings.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    templates_folder = f"{base_dir}/templates"
+
+    # Load strings package
+    strings_package_name = get_strings_package(base_dir)
+    strings_module = importlib.import_module(f"lite_content.{strings_package_name}.strings")
+
+    # Check all strings are found
+    lcs_strings = get_all_lcs_strings(templates_folder)
+    for string in lcs_strings:
+        if not find_occurance(strings_module, string.split(".")):
```
Comment (stephanieboyd) [inline templates/check_html_strings.py]: Can we call this `check_string_occurance`. Reads better in this condition, reads better that you're checking something

Comment (BenAndrew-Kainos) [reply templates/check_html_strings.py]: Done

File: templates/check_html_strings.py (hunk 8 above)
Comment (BenAndrew-Kainos) [reply templates/check_html_strings.py]: Done

**Pull request merged**